Feature/3 pacttype conversion trait

cowboy-bebug opened the pull request:
Initial attempt at "catch-all" trait.

A problem with `downcast_ref::<T>()` was that it requires a concrete type `T` to be specified, which meant we couldn't downcast to `AsRef<[u8]>` or `TryInto<u64>`.

Changes:
- add a new module (`src/interpreter/type_cast.rs`) with unit tests
- add necessary lines to integrate `type_cast` into the main library

Closes #3

cowboy-bebug committed:
add type_cast module

- add #![feature(specialization)] attribute at root level
- add AnyTryInto trait with any_try_into() method
- add default any_try_into() for T to return UnknownType err
- implement AnyTryInto for PactType

cowboy-bebug committed:
remove default impl & specialization, and add the rest of type conversion

cowboy-bebug committed:
fix build issue and refactor

- expose interpreter::type_cast module
- add cfg attribute
- specify correct crate
- expose AnyTryInto trait

jordy25519 commented on src/interpreter/mod.rs:
```diff
@@ -17,11 +17,14 @@
 pub mod types;
 use types::PactType;
 
+#[cfg(feature = "std")]
```
This needs to be `no_std` compatible, like the rest of the interpreter crate

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,170 @@
+use crate::interpreter::types::{Numeric, PactType, StringLike};
+use std::any::Any;
```
use `core::any` and `core::convert` to make it `no_std`

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,170 @@
+    fn any_try_into(value: &'a dyn Any) -> Result<PactType<'a>, PactConversionErr> {
+        // TODO: refactor the below repetiion using macros
+        // Signed integer type casting into PactType
+        if value.is::<i8>()
```
Is there a reason to do an `is` check before down casting?

cowboy-bebug replied:
Nope there is no need. All the `is` conditionals are removed

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,170 @@
+        // TODO: refactor the below repetiion using macros
+        // Signed integer type casting into PactType
+        if value.is::<i8>()
+            || value.is::<i16>()
```
I don't think we should support signed integers at this stage, since they aren't used in runtime methods IIRC

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,170 @@
+                }
+            }
+        }
+
```
Need to add casts for `Vec<u8>` and common crypto types like `[u8; 32]` and `[u8; 64]`

jordy25519 reviewed (changes requested):
Good progress so far

> A problem with downcast_ref::<T>() was that it requires a concrete type T to be specified, which meant we couldn't downcast to AsRef<[u8]> or TryInto<u64>.

We will have to change tact because of this. Since we can't use Any + traits to achieve this generically we lose the benefits of dynamic typing and since we're enumerating all the type mappings we might as well write individual `From<blah> for PactType<'_>` conversions in the first place, 

As mentioned in review comments:
- solution needs to be `no_std`
- `StringLike` conversions needed for common bytes types `(Vec<u8>, [u8; 32], [u8; 64])` maybe more that I've missed

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,170 @@
+    }
+}
+
+#[cfg(test)]
```
If a test like this can work, then there is still value in the `AnyTryInto` for associated types
```
trait Foo {
    type Number;
}

struct Bar;

impl Foo for Bar {
    type Number = u32;
}

let n = <Bar as Foo>::Number = 10u32;

assert_eq!(
    PactType::any_try_into(n)
    Ok(PactType::Numeric(Numeric(10)))
)
```

cowboy-bebug committed:
remove signed integer conversion

cowboy-bebug committed:
make no_std compatible

cowboy-bebug committed:
remove redundant  conditionals

cowboy-bebug committed:
add conversion for fixed hash types

- skip rustfmt for tests for readability
- add tests for associated type conversion
- improve tests

cowboy-bebug committed:
remove #[cfg(feature = std)] attribute for type_cast module

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,224 @@
+            if *number > core::u64::MAX as u128 {
+                return Err(PactConversionErr::Overflow);
+            }
+            if let Ok(n) = u64::try_from(*number) {
```
The `try_from` impl will check the overflow condition so no need to check here. Can map the error though into PactConversionError e.g. `try_from(*number).map_err(|_| PactConversionErr::Overflow)?`

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,224 @@
+        // TODO: refactor the below repetiion using macros
+        // Unsigned integer type casting into PactType
+        if let Some(number) = value.downcast_ref::<u8>() {
+            if let Ok(n) = u64::try_from(*number) {
```
It is safe and more efficient to use an `as` cast for `u8`, `u16`, and `u32` as they will never exceed `u64`

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    #[test]
+    fn it_converts_numeric_associated_types() {
```
🔥 🔥 🔥

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,224 @@
+use crate::interpreter::types::{Numeric, PactType, StringLike};
+use core::any::Any;
+use core::convert::TryFrom;
+use std::string::String;
```
Can use no_std variants here I think its under `alloc` instead of `std`

cowboy-bebug replied:
Yep it is using `alloc` which is imported as `std` 😀 
https://github.com/cennznet/pact/blob/master/src/lib.rs#L19

jordy25519 commented on src/interpreter/mod.rs:
```diff
@@ -14,14 +14,15 @@
 //!
 //! The pact bytecode interpreter
 //!
+pub mod type_cast;
```
Can we move this `mod` under `types` and export the `AnyTryInto` trait so that it is available through `interpreter::types::AnyTryInto`, now that we aren't gating `type_cast`

cowboy-bebug replied:
@holygits Adding `pub use crate::interpreter::type_cast::AnyTryInto;` in `src/interpreter/types.rs` results in unused import warning 😞 
^ is it okay?

jordy25519 replied:
Can add an `#[allow(unused_imports)]` to silence the warning

cowboy-bebug replied:
The warning is gone now

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,224 @@
+/// AnyTryInto implementation for PactType
+impl<'a> AnyTryInto<'a> for PactType<'a> {
+    fn any_try_into(value: &'a dyn Any) -> Result<PactType<'a>, PactConversionErr> {
+        // TODO: refactor the below repetiion using macros
```
typo: repetiion

jordy25519 commented on src/lib.rs:
```diff
@@ -26,6 +26,7 @@ extern crate pest;
 extern crate pest_derive;
 
 pub mod interpreter;
+pub use interpreter::type_cast;
```
Can remove this after moving `type_cast` module under `types` module

jordy25519 commented on src/interpreter/type_cast.rs:
```diff
@@ -0,0 +1,224 @@
+            type Number64 = u64;
+        }
+
+        let n32: <Bar as Foo>::Number32 = 10u32;
```
💭 Wondering does it work if we don't specify `10u32` i.e. just `10`?

cowboy-bebug replied:
Yep it does work because `10` gets coerced into `u32` (that's `Number32`).
But `(PactType::any_try_into(&10)` will produce `PactConversionErr::UnknownType` error because it defaults to `i32`, which we don't have implemented 😄

jordy25519 replied:
Perfect.

jordy25519 reviewed (changes requested):
Tests look really good. Few last changes requested

cowboy-bebug committed:
use 'as' casting and map_err() for overflow error

cowboy-bebug committed:
expose AnyTryInto trait from interpreter::types module
